### PR TITLE
*: Support user management through crd. #175

### LINKED ISSUE #175: [feature] Support user management through crd
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

Add, delete and modify users through user crd.

**Describe alternatives you've considered**

Extend kubectl. reference: https://github.com/kubernetes/sample-cli-plugin

**Additional context**

### PR DESCRIPTION
**Pull request by runkecheng**
### What type of PR is this?

/enhancement

### Which issue(s) this PR fixes?

Fixes #175 #198 #189 #220

### What this PR does?

Summary:

- [x] Support adding and deleting users
- [x] Support for user authorization (databases, tables)
- [x] Support password modification
- [x] Support modifying hosts
- [x] Support automatic installation of msyqluser crd
- [x] Add the verification of user parameters.

### Special notes for your reviewer?

The main logic is in this commit: [*: Support user management through crd. #175](https://github.com/radondb/radondb-mysql-kubernetes/pull/218/commits/1d7ed6584ec9fb769ff30c4436b4435d860bfe6f?file-filters%5B%5D=.go&manifests=true)

For test:
1. Delete the original cluster and crd, and redeploy using the test image.
    - operator image: `radondb/mysql-operator:test-2021910`
    - sidecar image: `radondb/mysql-sidecar:test-2021910`
2. Make sure that the root password of the msyql cluster is set.
3. Make sure to create user secret before creating user crd.

### CODE REVIEW

File: controllers/mysqluser_controller.go [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	mysqlv1alpha1 "github.com/radondb/radondb-mysql-kubernetes/api/v1alpha1"
```
Comment (andyli029) [inline controllers/mysqluser_controller.go:35]: mysqlv1alpha1->apiv1alpha1

File: controllers/mysqluser_controller.go [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+	"github.com/radondb/radondb-mysql-kubernetes/internal"
+	mysqluser "github.com/radondb/radondb-mysql-kubernetes/mysqluser"
```
Comment (andyli029) [inline controllers/mysqluser_controller.go:37]: Redundant alias.

File: controllers/mysqluser_controller.go [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+				return fmt.Errorf("set permissions faild, err: %s", err)
+			}
+		}
+	}
+
+	// 4. Delete the host that exists in the mysql but does not exist in the spec.
+	if err := r.updateAllowedHosts(ctx, sqlRunner, mysqlUser); err != nil {
+		return fmt.Errorf("update allowed hosts faild, err: %s", err)
+	}
+
```
Comment (acekingke) [inline controllers/mysqluser_controller.go:174]: delete empty line, it  belong to step 4.

File: internal/query.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+// Args representation of the query.
+func (q *Query) Args() []interface{} {
+	return q.args
+}
+
+// NewQuery returns a new Query object.
+func NewQuery(q string, args ...interface{}) Query {
+	if q == "" {
+		panic("unexpected empty query")
```
Comment (acekingke) [inline internal/query.go:41]: panic will exit the program， dangerous

File: controllers/mysqluser_controller.go [hunk 5]
```diff
@@ -0,0 +1,381 @@
+/*
```
Comment (andyli029) [inline controllers/mysqluser_controller.go:1]: Refactor the file.

File: internal/query.go (hunk 4 above)
Comment (andyli029) [reply internal/query.go]: Do not delete the logic `if q == ""` @runkecheng

File: controllers/mysqluser_controller.go [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	oldStatus := user.Status.DeepCopy()
+
+	sqlRunner, err := r.getSqlRunner(ctx, user)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("get sqlrunner faild, err: %s", err)
```
Comment (andyli029) [inline controllers/mysqluser_controller.go:82]: faild -> failed

File: controllers/mysqluser_controller.go [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+	for _, host := range mysqlUser.Spec.Hosts {
+		if err := checkUserExist(ctx, internal.CheckUserQuery(ctx, userName, host), sqlRunner); err != nil {
+			userLog.Info("mysql user not exist, create new mysql user", "need create host", host)
+			// unexistHosts save the hosts that need create.
+			unexistHosts = append(unexistHosts, host)
+		} else {
+			// 3. Set the password and permissions for the existing hosts according to the spec.
+			// Regardless of whether the password is modified or not, reset it to keep it up to date.
+			if err := r.setPassword(ctx, sqlRunner, userName, host, password); err != nil {
+				return fmt.Errorf("set password faild, err: %s", err)
```
Comment (andyli029) [inline controllers/mysqluser_controller.go:159]: Fix

File: controllers/mysqluser_controller.go [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+			}
+
+			host := []string{host}
+			// Regardless of whether the permissions is modified or not, reset it to keep it up to date.
+			if err := r.setPermission(ctx, sqlRunner, permissions, userName, host); err != nil {
+				return fmt.Errorf("set permissions faild, err: %s", err)
```
Comment (andyli029) [inline controllers/mysqluser_controller.go:165]: Fix

File: controllers/mysqluser_controller.go [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+
+	// 6. Add finalizer.
+	if err := r.addUserFinalizer(ctx, mysqlUser.Unwrap()); err != nil {
+		return fmt.Errorf("add finalizer faild")
+	}
+
+	// 7. Authorize newly created user hosts.
+	// Ensure that the user has been successfully created before authorization.
+	if err := r.setPermission(ctx, sqlRunner, permissions, userName, unexistHosts); err != nil {
+		return fmt.Errorf("set permissions faild, err: %s", err)
```
Comment (andyli029) [inline controllers/mysqluser_controller.go:194]: Fix

File: controllers/mysqluser_controller.go [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+		mysqluser.ConfigurationSucceededReason, "The mysql user has been configured successfully.",
+	)
+
+	return nil
+}
+
+// createUser create users in mysql based on cr information.
+func (r *MysqlUserReconciler) createUser(ctx context.Context, query internal.Query, sqlRunner *internal.SQLRunner) error {
+	if err := sqlRunner.RunQuery(query.String(), query.Args()...); err != nil {
+		return fmt.Errorf("create mysql user faild, err: %s", err)
```
Comment (andyli029) [inline controllers/mysqluser_controller.go:212]: Fix

File: controllers/mysqluser_controller.go [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+	defer sqlRunner.Close()
+
+	// If mysql user has been deleted then delete it from mysql cluster.
+	if !user.ObjectMeta.DeletionTimestamp.IsZero() {
+		queries := internal.GetDeleteQuery(ctx, user.Status.AllowedHosts, user.Spec.User)
+		return ctrl.Result{}, r.deleteAllUser(ctx, queries, sqlRunner, user)
+	}
+
+	// If mysql user present, create or update.
+	cuErr := r.createOrUpdate(ctx, sqlRunner, user)
```
Comment (andyli029) [inline controllers/mysqluser_controller.go:94]: Refactor it.

File: controllers/mysqluser_controller.go [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+// createOrUpdate create or update users in mysql.
+// 1. Check whether the necessary parameters exist and comply with the rules.
+// 2. Check if the spec.hosts exists in mysql.
+// 3. Set the password and permissions for the existing host according to the spec.
+// 4. Delete the host that exists in the mysql but does not exist in the spec.
+// 5. Create unexist hosts.
+// 6. Add finalizer.
+// 7. Authorize newly created user hosts.
+// 8. Update user`s status.
+func (r *MysqlUserReconciler) createOrUpdate(ctx context.Context, sqlRunner *internal.SQLRunner, mysqlUser *mysqluser.MysqlUser) error {
```
Comment (andyli029) [inline controllers/mysqluser_controller.go:117]: Refactor it.

**Pull request closed without merging**